### ISSUE TITLE
[Dynamo][Compile]Torch compile with dynamic shapes not working

### ISSUE THREAD

**Issue by bohnstingl** [pytorch/pytorch#105279]
### 🐛 Describe the bug

My networks rely on varying shapes during training as well as during inference. Thus, I tried to use `torch.compile(... dynamic=True)` as well as the `torch._dynamo.optimize(..., dynamic=True)` feature. However, unfortunately I can't get it to work properly and the function gets recompiled always as soon as the input shape changes. I tried it with the latest nightly build verison `2.1.0.dev20230712` as well as with the current main branch from `Friday Jul 14th 2023`. I printed the guards that trigger the recompilation and the culprit guard is precisely what the dynamic feature targets:
`GuardFail(reason="tensor 'L['input']' size mismatch at index 0. expected 146, actual 149", orig_code=<code object RNNScript at 0x2758840 ...)`

I also tried to add a `warm-up phase`, where the input length is varied, but whenever the shape of the input changes, the function gets recompiled.

@ezyang could you maybe please advice here?

### Error logs

GuardFail(reason="tensor 'L['input']' size mismatch at index 0. expected 146, actual 149", orig_code=<code object RNNScript at 0x2758840 ...)

### Minified repro

```
from typing import List, Tuple, Optional, overload, Union, cast
import torch
import numpy as np
import time
import torch.optim as optim
from torch.nn.parameter import Parameter

def RNNScript(
    input,
    param1,
    param2,
    ):

    state1 = torch.zeros(32, 340, dtype=input.dtype, device=input.device)
    
    outs = []

    Wx = input @ param1
    Wx_inp, Wx_rec = torch.tensor_split(Wx, 2, 2)
    for wt_inp, wt_rec in zip(Wx_inp, Wx_rec):
        rec_mul_inp, rec_mul_rec = torch.tensor_split(state1 @ param2, 2, 1)
        input_prev = (wt_inp + rec_mul_inp)
        output_gate = (wt_rec + rec_mul_rec)

        state1 = input_prev * torch.sigmoid(output_gate)
        outs.append(state1)
    
    outs = torch.stack(outs)

    return outs, (outs)

if __name__ == "__main__":

    input_size = 140
    hidden_size = 340
    num_layers = 1
    num_timesteps = 111
    batch_size = 32

    bi_dir = True 
    rnnt_input = False
    num_threads = -1
    use_gpu = True
    load_weights = False

    forward_times = []
    backward_times = []

    if use_gpu:
        device = torch.device('cuda:0')
    else:
        device = None

    parameters = []

    w_ih = torch.empty((input_size, hidden_size), device=device)
    w_io = torch.empty((input_size, hidden_size), device=device)
    w_i_comb = Parameter(torch.cat([w_ih,w_io],1))
    parameters.append(w_i_comb)

    w_hh = torch.empty((hidden_size, hidden_size), device=device)
    w_ho = torch.empty((hidden_size, hidden_size), device=device)
    w_h_comb = Parameter(torch.cat([w_hh,w_ho],1))
    parameters.append(w_h_comb)
    
    def count_kernels(guard):
        print("[pt2_compile] guard failed: ", guard)

    rnnscript = torch.compile(RNNScript, mode='reduce-overhead', dynamic=True, fullgraph=True)
    #backend = torch._TorchCompileInductorWrapper('reduce-overhead', None, True)
    #rnnscript = torch._dynamo.optimize(backend=backend, nopython=True, dynamic=True, guard_fail_fn=count_kernels)(RNNScript)
    #rnnscript = RNNScript
    snu = lambda x: rnnscript(x, w_i_comb, w_h_comb)

    optimizer = optim.SGD(parameters, 0.1)

    inp = torch.rand((num_timesteps, batch_size, input_size))

    if use_gpu:
        inp = inp.cuda()

    optimizer.zero_grad()
    for execution in range(5):
        start_forward = time.time_ns()
        t_rnd = np.random.randint(0, 200)
        inp = torch.rand((t_rnd, batch_size, input_size))
        if use_gpu:
            inp = inp.cuda()
        out, state = snu(inp)

        if use_gpu:
            torch.cuda.synchronize()
        stop_forward = time.time_ns()
        forward_times.append((stop_forward - start_forward) / (10 ** 9))

        loss = 1. - torch.sum(out)

        start_time_backward = time.time_ns()
        #loss.backward()
        if use_gpu:
            torch.cuda.synchronize()
        stop_time_backward = time.time_ns()
        backward_times.append((stop_time_backward - start_time_backward) / (10 ** 9))

    print('================================================================')
    print('Model with sSNU-os:')
    print('# Layers: ' + str(num_layers))
    print('# Units per layer: ' + str(hidden_size))
    print('Bidirectional: ' + str(bi_dir))
    print('Load weights: '  + str(load_weights))
    print('RNN-T input: ' + str(rnnt_input))
    print('# CPU threads: ' + str(num_threads))
    print('GPU support: ' + str(use_gpu))
    print('----------------------------------------------------------------')
    print('Timing summary')
    print('Time of forward computation: {:.4f} +- {:.4f} s'.format(np.mean(np.array(forward_times)), np.std(np.array(forward_times))))
    print('Time of backward computation: {:.4f} +- {:.4f} s'.format(np.mean(np.array(backward_times)), np.std(np.array(backward_times))))

```

### Versions

Collecting environment information...
PyTorch version: 2.1.0a0+gitfb376f8
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.7 (Ootpa) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-16)
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.28

Python version: 3.11.3 (main, May 15 2023, 15:45:52) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-425.10.1.el8_7.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 525.60.13
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              128
On-line CPU(s) list: 0-127
Thread(s) per core:  1
Core(s) per socket:  64
Socket(s):           2
NUMA node(s):        2
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7742 64-Core Processor
Stepping:            0
CPU MHz:             3361.974
CPU max MHz:         2250.0000
CPU min MHz:         1500.0000
BogoMIPS:            4499.91
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-63
NUMA node1 CPU(s):   64-127
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] pytorch-triton==2.1.0+3c400e7818
[pip3] torch==2.1.0a0+gitfb376f8
[pip3] torchaudio==2.1.0a0+cf53a48
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.8.0               h6a678d5_0  
[conda] magma-cuda118             2.6.1                         1    pytorch
[conda] mkl                       2023.1.0         h6d00ec8_46342  
[conda] mkl-include               2023.1.0         h06a4308_46342  
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl_fft                   1.3.6           py311ha02d727_1  
[conda] mkl_random                1.2.2           py311ha02d727_1  
[conda] numpy                     1.24.3          py311h08b1b3b_1  
[conda] numpy-base                1.24.3          py311hf175353_1  
[conda] pytorch-triton            2.1.0+3c400e7818          pypi_0    pypi
[conda] torch                     2.1.0a0+gitfb376f8          pypi_0    pypi
[conda] torchaudio                2.1.0a0+cf53a48           dev_0    <develop>

**Comment by ezyang** [comment]
When I run your example with TORCH_LOGS=dynamic I get

```
[2023-07-15 18:09:58,188] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_env
[2023-07-15 18:09:58,480] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_symbol s0 = 14 for L['input'].size()[0]
[2023-07-15 18:09:58,517] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_symbol s1 = 32 for L['input'].size()[1]
[2023-07-15 18:09:58,519] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_symbol s2 = 140 for L['input'].size()[2]
[2023-07-15 18:09:58,601] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: eval Eq(s2, 140) [guard added] at rn.py:18 in RNNScript (_meta_registrations.py:1650 in meta_mm)
[2023-07-15 18:09:58,626] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: eval Eq(s0, 14) [guard added] at rn.py:20 in RNNScript (_dynamo/variables/tensor.py:651 in evaluate_expr)
[2023-07-15 18:09:58,687] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: eval Eq(s1, 32) [guard added] at rn.py:22 in RNNScript (_subclasses/fake_tensor.py:674 in infer_size)
```

There are few things going on here.

The specialization at line 18 is due to a matrix multiply with a Parameter, which we assume is always static. Since you're varying parameter sizes, this assumption is not good for you. You can un-Parameter'ify these tensors, but we should also add a knob that overrides parameter specialization.

The specialization at line 20 is more difficult to deal with. It comes from

```
    for wt_inp, wt_rec in zip(Wx_inp, Wx_rec):
```

In other words, you are iterating over the batch dimension of Wx_inp/Wx_rec. This will necessarily specialize, because we always have to specialize loop iteration counts to do loops in Python, since our strategy is always to unroll them. To avoid specialization here, you'll have to rewrite your code without loops. Vmap can help write your code in the single batch style without an explicit loop (though cc @zou3519, I don't remember if you have to do anything extra to make it work in torch.compile.)

**Comment by bohnstingl** [comment]
Thank you so much for your very quick and detailed insights here! They are very helpful!

When you say `un-Parameter'ify`, you mean using `Tensor(..., require_grad=True)` instead of a `Parameter(...)`? 

As you might have noticed, my code should emulate the function of a simple RNN unit. In the static case this works perfectly fine, but the dynamic case is giving me trouble. Therefore, `for wt_inp, wt_rec in zip(Wx_inp, Wx_rec):` iterates over the `time dimension` of the multiplication result (The shape of the input is `time x batch x features`). I would be more than happy to learn to rewrite my code in a way that it works with `torch.compile(..., dynamic=True)` as I think this features is just amazing.

**Comment by vadimkantorov** [comment]
@ezyang btw this might me related to the torch.scan/loop discussion. If we get one dynamic variant to work, it could be an example or even torch.scan could be implemented using just a loop (if we have tests that it compiles into sth nice). Custom RNNs is a prototypical use case for this...

**Comment by bohnstingl** [comment]
@ezyang I have modified my example using `torch.Tensor` instead of `Parameter` and the guards that are added in that case look fine to me.

However, the for-loop is something that I cannot work around.
I tried all sorts of variations for the for-loop as well as a while loops with branches, but the guards that are added are always fully constraint to the specific time dimension of the tensor. @vadimkantorov do you have an example of the `torch.scan/loop` where a dynamic number of iterations is supported?

**Comment by vadimkantorov** [comment]
No working example, just the discussion here: https://github.com/pytorch/pytorch/issues/50688

If your use case can be implemented by efficiently by the relevant functions in jax, I would say it's one more argument to have similar functions in pytorch

**Comment by bohnstingl** [comment]
I, see. With Jax, custom RNNs should be simple to implement using `jax.lax.scan` as illustrated [here](https://svenschmit.com/jax-language-model-rnn). Also this was done efficiently in TF, I believe. To me in general having the capability to evaluate a function on a per-timestep basis would be important. This is also similar to #103588, I think.

As I mentioned, `torch.compile()`, even with the dynamic mode, introduces guards that trigger a recompilation of the custom RNN in PyTorch. So the for me dynamic input lengths are not working.

**Comment by bohnstingl** [comment]
Could this pull request #104442 resolve this issue?

**Comment by bohnstingl** [comment]
@ezyang I just wanted to follow-up on the topic of the parameter specialization you mentioned in case of the matrix multiplication. Is there a plan to change this, such that the the matmul with Parameters behaves similarly as the matmul with Tensors as far as the compilation is concerned?

I don't have a good overview of the PT2 parts, but if there is something I can help with, I would be happy to

**Comment by ezyang** [comment]
Look at `tensor_always_has_static_shape` in `torch/_dynamo/utils.py`. Add a config to bypass this and always return False, None. Send a PR and add me as reviewer. Should be fairly simple. Don't forget to add a test.

**Comment by bohnstingl** [comment]
@ezyang Since parameter part of this issue is in the making, is there a way to address the second part of this issue?
I looked at `torch.vmap`, but I wouldn't know how to implement RNNs with it, because if one wants to use it to run a function over the time dimension, it does not allow for a carry, right? Something like scan would then be needed, which has demanded also a lot, see [here](https://github.com/pytorch/pytorch/issues/50688) and [here](https://github.com/pytorch/pytorch/pull/1044420).

Alternatively to that, would it be possible that the python for-loop is not always unrolled, but translated into a for loop inside the triton kernel? This might be similarly effective as the unrolling, but more flexible to allow dynamic shapes?

**Comment by ezyang** [comment]
If you want to do an device scan you will have to wait for @peterbell10 to get to it alas

**Comment by bohnstingl** [comment]
It is not necessarily a device scan operation. It would suffice already if the for-loop is not unrolled fully, but rather translated down into a for-loop in triton. This should allow for the case where the loop is over a dimension of the tensor and this shape is dynamic, correct? Is this very difficult to do? 

@peterbell10 I am more than happy to help as much as I can on this.

**Comment by bohnstingl** [comment]
@ezyang Naggy the issue with the `for-loop` supporting dynamic shapes a bit. Is there something that I can help with?
Down below I have provided a very short test script that showcases the issue that I am talking about.

```
from typing import List, Tuple, Optional, overload, Union, cast
import torch
import numpy as np
import time
import torch.optim as optim
from torch.nn.parameter import Parameter

class RNNTest():

    def __init__(self, device) -> None:
        self.w = torch.empty((input_size, hidden_size), device=device)
        self.w.requires_grad_(True)

        self.w_h = torch.empty((hidden_size, hidden_size), device=device)
        self.w_h.requires_grad_(True)
    
    def RNNScript(self, input):

        state1 = torch.zeros(batch_size, hidden_size, dtype=input.dtype, device=input.device)
        outs = []

        Wx = input @ self.w
        for wt in Wx:
            rec_mul = state1 @ self.w_h
            input_prev = (wt + rec_mul)

            state1 = 1. + torch.sigmoid(input_prev)
            outs.append(state1)
        
        outs = torch.stack(outs)

        return outs, None

if __name__ == "__main__":

    input_size = 140
    hidden_size = 340
    batch_size = 64
    use_gpu = True

    if use_gpu:
        device = torch.device('cuda:0')
    else:
        device = None

    rnn_test = RNNTest(device)
    
    def count_kernels(guard):
        print("[pt2_compile] guard failed: ", guard)

    rnnscript = torch.compile(rnn_test.RNNScript, dynamic=True)
    #rnnscript = rnn_test.RNNScript
    
    optimizer = optim.SGD([rnn_test.w, rnn_test.w_h], 0.1)
    optimizer.zero_grad()

    for execution in range(20):
        t_rnd = 20
        inp = torch.rand((t_rnd, batch_size, input_size))
        print(inp.shape)
        if use_gpu:
            inp = inp.cuda()
        
        out, state = rnnscript(inp)

        if use_gpu:
            torch.cuda.synchronize()
        
        loss = 1. - torch.sum(out)
        loss.backward()

        if use_gpu:
            torch.cuda.synchronize()
        
```

```
[2023-08-02 14:22:20,294] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_env
[2023-08-02 14:22:20,319] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_symbol s0 = 20 for L['input'].size()[0]
[2023-08-02 14:22:20,350] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_symbol s1 = 64 for L['input'].size()[1]
[2023-08-02 14:22:20,352] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_symbol s2 = 140 for L['input'].size()[2]
[2023-08-02 14:22:20,371] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: create_symbol s3 = 340 for L['self'].w.size()[1]
[2023-08-02 14:22:20,414] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: eval Eq(s0, 20) [guard added] at u/gsaboh/RNN-T_git/pylib/PT2_testfile8.py:23 in RNNScript (_dynamo/variables/tensor.py:704 in evaluate_expr)
[2023-08-02 14:22:20,520] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: eval Eq(s3, 340) [guard added] at u/gsaboh/RNN-T_git/pylib/PT2_testfile8.py:24 in RNNScript (_meta_registrations.py:1604 in meta_mm)
[2023-08-02 14:22:20,526] torch.fx.experimental.symbolic_shapes: [INFO] 0.0: eval Eq(s1, 64) [guard added] at u/gsaboh/RNN-T_git/pylib/PT2_testfile8.py:25 in RNNScript (_subclasses/fake_tensor.py:711 in infer_size)
```

The critical line is the guard with `Eq(s0, 20)`, which is the subject of this discussion.

**Comment by bohnstingl** [comment]
@ezyang @peterbell10 Can anybody give me some guidance on where to start looking into this issue? For example, where is the unrolling of the for-loop implemented and where can I replace this unrolling with a for-loop in triton?

**Comment by bohnstingl** [comment]
Nagging this issue again.
There is the open feature request for a [scan functionality](https://github.com/pytorch/pytorch/pull/104442), but as far as I can see its implementation also contains a for-loop, which will later enforce the respective dimension to be static, right?

@ezyang @peterbell10, can help me with this issue, or do you know of somebody who can?

**Comment by ezyang** [comment]
@peterbell10 landed inductor scan support. We just need the HOP, which I think Peter said he was also going to hit.

**Comment by bohnstingl** [comment]
This is awesome, thank you @peterbell10. I had a look at the code and I believe currently scan has inductor support for `cumsum` and `cumprod`, correct? Are there plans to also support arbitrary functions, is this the `HOP` support that you mentioned?

**Comment by peterbell10** [comment]
Unfortunately  support for `cumsum` and `cumprod` was reverted due to triton bugs but I'm currently working on updating our pinned triton version so we can reland that PR. See https://github.com/pytorch/pytorch/pull/106581#issuecomment-1719811113

I also have a work in progress PR that generates arbitrary single-tensor scans in triton from inductor IR which is the first step towards a higher order `torch.scan` operator.

**Comment by bohnstingl** [comment]
I see, thank you for this information! I will definitely follow this PR and your commits.

Regarding the generic scan functionality, is there something that I can be of help?

**Comment by peterbell10** [comment]
Hrm I just realized the full scope of what's being asked for here. triton's scan operators work on scalars and require the operator to be associative. However, here we are scanning over matrices of values and the operator doesn't look associative to me.

Without associativity, the iterations need to be executed serially and the best we could do here might just be generating a loop in the wrapper code that invokes the same kernel a dynamic number of times.

**Comment by bohnstingl** [comment]
Thank you @peterbell10 for looking into this. You are absolutely correct, the problem here is that the operator (function) is not associative and hence cannot rely on triton's associative_scan. 
What you describe is what I was having in mind actually. For loops are supported in triton, correct? Thus one could have a for loop and a triton compile body. This should still be more performant as compared to a for loop in Pytorch and just the body being compiled, right?

Maybe something like this
```
@triton.jit
def _loop_body(inp, old_state):
    <Code of loop body>
    return new_out, new_state

@triton.jit
def triton_(in_ptr0, out_ptr0, xnumel, rnumel, XBLOCK : tl.constexpr):
    <Boilerplate; load initial out>
    for xoffset in range(maxT):
        out, state = _loop_body(tl.load(in_ptr0 + xoffset), state)
    tl.store(out_ptr0), out)
```

I know that the above is not operational code, but I hope it represents the essence. I am wondering, in the generic scan function the operator is not generally associative, right? So would it be possible to have flag or something in the scan operator that indicates if the operation is associative? If yes, then use triton's associative_scan, if no then use a for loop to iterate over the respective tensor dimension with a for loop and just compile the loop-body, i.e., the operator? Could one integrate this into https://github.com/pytorch/pytorch/pull/109132 maybe?

**Comment by bohnstingl** [comment]
I tried to look at this problem for some time and the main problem that I can see is that for loops are unrolled, which prevents the use of dynamic shapes. So one somehow needs to lower the for loop to the inductor, which I though could be done via the scan functionality?

**Comment by peterbell10** [comment]
> For loops are supported in triton, correct? Thus one could have a for loop and a triton compile body. This should still be more performant as compared to a for loop in Pytorch and just the body being compiled, right?

Our matmul kernels are currently templates so wouldn't be flexible enough to generate this code (would be possible with very significant changes). For simpler cases like pointwise operations it should be reasonably straightforward

Also in general, we can't implement all operators in triton so we would need to fallback to a python loop at least in some cases.

> So would it be possible to have flag or something in the scan operator that indicates if the operation is associative?

The signature of `lax.scan` which is copied in #104442 would not be suitable for an associative scan, so I think they would need to be distinct operators.

> I tried to look at this problem for some time and the main problem that I can see is that for loops are unrolled, which prevents the use of dynamic shapes. So one somehow needs to lower the for loop to the inductor, which I though could be done via the scan functionality?

Yes I think the scan interface defined in #104442 would be suitable for this use case.

**Comment by bohnstingl** [comment]
> Our matmul kernels are currently templates so wouldn't be flexible enough to generate this code (would be possible with very significant changes). For simpler cases like pointwise operations it should be reasonably straightforward
> 
> Also in general, we can't implement all operators in triton so we would need to fallback to a python loop at least in some cases.

I don't quite get this. In my very initial example, the loop body should be compilable, even with dynamic shapes, correct? The only issue that breaks the dynamic shapes at the moment is the unrolling of the for loop on the PyTorch level. Thus, I thought that if the loop body gets compiled to triton code and the for loop is placed on the inductor level, then dynamic shapes should work?

> The signature of lax.scan which is copied in https://github.com/pytorch/pytorch/pull/104442 would not be suitable for an associative scan, so I think they would need to be distinct operators.
> 
> [...]
> 
> Yes I think the scan interface defined in https://github.com/pytorch/pytorch/pull/104442 would be suitable for this use case.

So if I may summarize: The scan interface defined in https://github.com/pytorch/pytorch/pull/104442 would be suitable for the non-associative scan functionality that would be needed here, but it does not yet have inductor lowering support. On the other hand, the interface that you outlined in https://github.com/pytorch/pytorch/pull/109132 has inductor lowering, but is mainly for associative scan? 

What would be your suggestion for a course of action here? Use the interface in https://github.com/pytorch/pytorch/pull/104442 and add inductor lowering, or create a separate interface based on https://github.com/pytorch/pytorch/pull/109132 which would support non-associative operations?

**Comment by peterbell10** [comment]
>  I thought that if the loop body gets compiled to triton code and the for loop is placed on the inductor level, then dynamic shapes should work?

Yes if the loop represented explicitly at the inductor level then we can compile this code without flattening the loop. We could always generate a python loop equivalent to the original code but inside inductor's wrapper code and this fixes the dynamic loop size issue. 

However, writing those loops inside the triton kernel is non-trivial, since each CUDA block runs its own loop and may be executing different parts of the code at any given moment so you need to prove there are no data dependencies between CUDA blocks. Inductor also doesn't implement everything in triton, so if anything that can't be lowered to triton appears in your combine function then we can't generate a triton loop.

> What would be your suggestion for a course of action here? Use the interface in https://github.com/pytorch/pytorch/pull/104442 and add inductor lowering, or create a separate interface based on https://github.com/pytorch/pytorch/pull/109132 which would support non-associative operations?

Lowering #104442 to inductor would be the best starting point here I believe (though I haven't looked at the code too closely).

**Comment by vadimkantorov** [comment]
> Without associativity, the iterations need to be executed serially and the best we could do here might just be generating a loop in the wrapper code that invokes the same kernel a dynamic number of times.

I know that sometimes people produce kernels with a few iterations of the underlying RNN cell. This way, fewer kernels launches are needed and hence less overhead. For some static shapes or a short sequences, maybe just an unrolled kernel can be generated?
